src/librbd/plugin/Types.h

```diff
@@ -14,8 +15,43 @@ namespace plugin {
 
 template <typename> struct Api;
 
+template <typename ImageCtxT>
```

Comment:
Nit: I don't think this needs to be templated. You can create a derived version under `plugin::WriteLogImageCache::HookPoints` which will have access to `plugin::WriteLogImageCache`'s privates (including the `ImageCtx`).

Reply:
Delete HookPointsInterface, and created a new class plugin::WriteLogImageCache::HookPoints.

Reply:
One more thing, such instances plugin::WriteLogImageCache and plugin::ParentCache are shared for multiple images, it should not contain the field m_image_ctx. So I create a PR to remove it: https://github.com/ceph/ceph/pull/38201